### PR TITLE
Bandit as a dynamic variant

### PR DESCRIPTION
**Pull request by aarsilv**
* **Ticket:** [FF-1007 - Adjust Beta Java SDK to handle dynamic variant and updated logging strategy](https://linear.app/eppo/issue/FF-1007/adjust-beta-java-sdk-to-handle-dynamic-variant-and-updated-logging)
* **Design Review:** [Bandits Engineering Design](https://www.notion.so/eppo/Bandits-Engineering-Design-c5df5b6072b446aa8cacbf54e4d1f9df?pvs=4) (Relevant sections: [RAC](https://www.notion.so/eppo/Bandits-Engineering-Design-c5df5b6072b446aa8cacbf54e4d1f9df?pvs=4#1f137a8a91cc42aebca4d5aea3c49630), [The Customer's Logger](https://www.notion.so/eppo/Bandits-Engineering-Design-c5df5b6072b446aa8cacbf54e4d1f9df?pvs=4#3207458fb01649c187d0bd3145bd0ea5))

## What Changed

These changes continue to evolve our beta/unused prototype Java SDK to support bandit experiments. The end goal is once this fork is in a good spot, we'd merge it back into our main SDK.

The two key changes from this PR are:

1. Bandits are now variation-level "dynamic variants." This allows them to be baked into experiments from the get-go and--in the future--possibly pitted against other bandits. A "dynamic variant" is a variant whose action is not constant. While bandits are the only dynamic variants right now, in the future, we could have other dynamic variant types (e.g., Eppo recommendation engine, pick a totally random action, etc.)

2. We provide a way to attach a bandit logger to the client that will log bandit-specific stuff (e.g., selected action, probability of that action, action attributes, model version, etc.)

3. We provide a vehicle to log an action the customer selects (as opposed to the bandit). This is useful for enabling complete transparency reports on what users have seen and may help power our learnings, too. 

## Change Details

The core change to support dynamic variants and customer action logging was to pull out the logic for determining the assigned variation into its own private helper method. That way, we could retrieve the variation at play when determining the variation and also later when logging a non-bandit action. The latter is useful so that we have key log information like the allocation key.

Since we log lots of good information (e.g., allocation key, experiment key, probability--computed from subject shards, etc.) the helper function returns a new POJO, `VariatonAssignmentResult`, used to pass all the intermediately computed information around so it can later be logged.

For the logger itself, we introduced the `IBanditLogger` interface, and `BanditLogData` POJO.

## Looking ahead

The next step is to create a v3 interface that returns a complex object. `VariatonAssignmentResult` may be seen as an early version of this.

Note: LaunchDarkly does this with their `getXXXVariationDetail()` methods ([example](https://launchdarkly.github.io/java-server-sdk/com/launchdarkly/sdk/server/LDClient.html#stringVariationDetail(java.lang.String,com.launchdarkly.sdk.LDContext,java.lang.String))) that return an [`EvaluationDetail`](https://launchdarkly.github.io/java-server-sdk/com/launchdarkly/sdk/EvaluationDetail.html) object.

## Example potential customer Java code

```java
//=== CLIENT SETUP ===//

EppoClientConfig config = EppoClientConfig.builder()
  .apiKey("abcdef1234567890")
  .assignmentLogger(snowflakeAssignmentLogger)
  .banditLogger(snowflakeBanditLogger)
  .build();
  
EppoClient.init(config);

//=== INPUT ===//

// Experiment's feature flag key, from our UI/UX
String flagKey = "ad bandit";

// Subject information, provided by customer
String subjectKey = "user12345";
EppoAttributes subjectAttributes = new EppoAttributes();
subjectAttributes.put("gender identity", EppoValue.of("female"));
subjectAttributes.put("nike brand affinity", EppoValue.of(0.6));
subjectAttributes.put("adidas brand affinity", EppoValue.of(0.2));

// Possible actions and their attributes, provided by customer
Map<String, EppoAttributes> possibleActionsAndAttributes = new Map<>();

String nikeAdAction = "nike ad";
EppoAttributes nikeAdAttributes = new EppoAttributes();
nikeAdAttributes.put("brand", "nike");
nikeAdAttributes.put("model gender", "male");
possibleActions.put(nikeAction, nikeAdAttributes);

String adidasAdAction = "adidas ad";
EppoAttributes adidasAdAttributes = new EppoAttributes();
adidasAdAttributes.put("brand", "nike");
adidasAdAttributes.put("model gender", "male");
possibleActions.put(nikeAction, nikeAdAttributes);

String pumaAdAction = "puma ad";
EppoAttributes pumaAdAttributes = new EppoAttributes();
adidasAdAttributes.put("brand", "puma");
possibleActions.put(pumaAdAction, adidasAdAttributes);

//=== BANDIT CALL ===//

String action;

// Get the assignment
Optional<String> stringAssignment = EppoClient.getInstance().getStringAssignment(
  subjectKey,
  flagKey,
  subjectAttributes,
  possibleActions
);

if (stringAssignment != null && stringAssignment.get() != "control") {
  // Bandit chose an action!
  action = stringAssignment.get();
} else {
  // Handle error/control situation where bandit is not at play
  action = determineActionFromRecommendationEngine(subjectKey, subjectAttributes, possibleActions);

  if (stringAssignment != null && stringAssignment.get() == "control") {
   // Not an error situation, but control intentionally selected 
   EppoClient.getInstance().logNonBanditAction(
      subjectKey,
      flagKey,
      subjectAttributes,
      action,
      possibleActions.get(action)
    );
  }
}

//=== ACTION USE ===//

// Get ad info to send to front-end
Ad adToShow = getAdInformation(action);
return adToShow;
```

### CODE REVIEW

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 1]
```diff
@@ -68,96 +52,154 @@ private Optional<EppoValue> getAssignmentValue(
         InputValidator.validateNotBlank(subjectKey, "Invalid argument: subjectKey cannot be blank");
         InputValidator.validateNotBlank(flagKey, "Invalid argument: flagKey cannot be blank");
 
+        VariationAssignmentResult assignmentResult = this.getAssignedVariation(flagKey, subjectKey, subjectAttributes);
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:54]: This is now using the refactored-out method to get a `VariationAssignmentResult` that includes the assigned variant (or `null`, if none) and useful intermediate computed information.

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 2]
```diff
@@ -68,96 +52,154 @@ private Optional<EppoValue> getAssignmentValue(
...
+        Optional<EppoValue> assignmentValue = Optional.of(assignedVariation.getTypedValue());
+
+        // Below is used for logging
+        String experimentKey = assignmentResult.getExperimentKey();
+        String allocationKey = assignmentResult.getAllocationKey();
+        String assignedVariationString = assignedVariation.getTypedValue().stringValue();
+
+        if (assignedVariation.getAlgorithmType() == AlgorithmType.OVERRIDE) {
+            // Assigned variation was from an override; return its value without logging
+            return assignmentValue;
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:70]: For now, I'll keep supporting overrides as long as the RAC generation code supports them (even if it's not in our UI/UX). Seems like the safe thing to do.

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 3]
```diff
@@ -68,96 +52,154 @@ private Optional<EppoValue> getAssignmentValue(
...
+                            subjectKey,
+                            subjectAttributes));
+        } catch (Exception e) {
+            log.warn("Error logging assignment", e);
+        }
+
+        return assignmentValue;
+    }
+
+    private VariationAssignmentResult getAssignedVariation(String flagKey, String subjectKey, EppoAttributes subjectAttributes) {
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:93]: Start of new reusable helper method that most code was moved into.

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 4]
```diff
@@ -68,96 +52,154 @@ private Optional<EppoValue> getAssignmentValue(
...
+        String experimentKey = assignmentResult.getExperimentKey();
+        String allocationKey = assignmentResult.getAllocationKey();
+        String assignedVariationString = assignedVariation.getTypedValue().stringValue();
+
+        if (assignedVariation.getAlgorithmType() == AlgorithmType.OVERRIDE) {
+            // Assigned variation was from an override; return its value without logging
+            return assignmentValue;
+        } else if (assignedVariation.getAlgorithmType() == AlgorithmType.BANDIT) {
+            // Assigned variation is a bandit; need to use the bandit to determine its value
+            assignmentValue = this.determineAndLogBanditAction(assignmentResult, assignmentOptions);
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:73]: Bandit-specific stuff also consolidated to a private helper method.

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 5]
```diff
@@ -68,96 +52,154 @@ private Optional<EppoValue> getAssignmentValue(
...
+    }
 
-            // Check if in experiment sample
-            allocationKey = rule.get().getAllocationKey();
-            Allocation allocation = configuration.getAllocation(allocationKey);
-            if (!this.isInExperimentSample(subjectKey, flagKey, configuration.getSubjectShards(),
-              allocation.getPercentExposure())) {
-                log.info("[Eppo SDK] No assigned variation. The subject is not part of the sample population");
-                return Optional.empty();
+    private Optional<EppoValue> determineAndLogBanditAction(VariationAssignmentResult assignmentResult, Map<String, EppoAttributes> assignmentOptions) {
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:154]: I know having the word "and" in a method name is a code smell ('it does two things!') but the theoretically cleaner way of having a separate `logBanditAction()` method would--as with variant selection--require passing around a ton of intermediate computed information and metadata. So then we'd need a `BanditActionResult` class. This dissuaded me from doing it, although it may indeed still be the better way.

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 6]
```diff
@@ -449,6 +467,50 @@ private EppoValue getSubjectVariationOverride(
...
+     * Logs an action taken that was not selected by the bandit.
+     * Useful for full transparency on what users experienced.
+     * @param subjectKey subjectKey identifier of the experiment subject, for example a user ID.
+     * @param flagKey feature flag, bandit, or experiment identifier
+     * @param subjectAttributes optional attributes associated with the subject, for example name, email, account age, etc.
+     * @param actionString name of the action taken for the subject
+     * @param actionAttributes attributes associated with the given action
+     * @return null if no exception was encountered by logging; otherwise, the encountered exception
+     */
+    public Exception logNonBanditAction(
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/EppoClient.java:479]: For now, in lieu of a whole complex logging result object, I'll just return any exceptions encountered so customers can deal with them. I imagine in v3, this will return a logging result of sorts.

File: src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java [hunk 7]
```diff
@@ -28,45 +25,32 @@ public static List<Variation> evaluateBanditVariations(
...
-        }
+    private static int hashToPositiveInt(String experimentKey, String subjectKey, String actionKey) {
+        int SHUFFLE_SHARDS = 10000;
+        return Shard.getShard(experimentKey+"-"+subjectKey+"-"+actionKey, SHUFFLE_SHARDS);
     }
 
     private static List<Variation> generateVariations(List<String> shuffledActions, Map<String, Float>  actionWeights, int subjectShards) {
 
-        final AtomicInteger lastShard = new AtomicInteger(-1);
+        final AtomicInteger lastShard = new AtomicInteger(0);
```
Comment (aarsilv) [inline src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java:47]: Didn't realize end shard was _exclusive_

File: src/main/java/com/eppo/sdk/helpers/Shard.java [hunk 8]
```diff
@@ -42,9 +42,9 @@ public static String getHex(String input) {
      * @return
      */
     public static int getShard(String input, int maxShardValue) {
-        String hashText = Shard.getHex(input);
+        StringBuilder hashText = new StringBuilder(Shard.getHex(input));
         while (hashText.length() < 32) {
-            hashText = "0" + hashText;
+            hashText.insert(0, "0");
```
Comment (leoromanovsky) [inline src/main/java/com/eppo/sdk/helpers/Shard.java:47]: 👍 if you're going to write Java, do it the Java way 😄

File: src/test/resources/bandits/rac-experiments-bandits-beta.json [hunk 9]
```diff
@@ -1,14 +1,42 @@
...
+            },
+            {
+              "name": "bandit",
+              "value": "banner-bandit",
+              "typedValue": "banner-bandit",
+              "shardRange": {
+                "start": 2000,
+                "end": 10000
+              },
+              "algorithmType": "bandit"
```
Comment (leoromanovsky) [inline src/test/resources/bandits/rac-experiments-bandits-beta.json:35]: Will `algorithmType` key be added to all variations?

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 10]
```diff
@@ -68,96 +51,153 @@ private Optional<EppoValue> getAssignmentValue(
...
+    }
+
+    private VariationAssignmentResult getAssignedVariation(String flagKey, String subjectKey, EppoAttributes subjectAttributes) {
+
         // Fetch Experiment Configuration
         ExperimentConfiguration configuration = this.configurationStore.getExperimentConfiguration(flagKey);
         if (configuration == null) {
             log.warn("[Eppo SDK] No configuration found for key: " + flagKey);
-            return Optional.empty();
+            return null;
```
Comment (leoromanovsky) [inline src/main/java/com/eppo/sdk/EppoClient.java:99]: Why the change to returning `null` ?

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 11]
```diff
@@ -449,6 +466,50 @@ private EppoValue getSubjectVariationOverride(
...
+        Exception loggingException = null;
+        try {
+            VariationAssignmentResult assignmentResult = this.getAssignedVariation(flagKey, subjectKey, subjectAttributes);
+
+            if (assignmentResult == null) {
+                // No bandit at play
+                return null;
+            }
+
+            String variationValue = assignmentResult.getVariation().getTypedValue().stringValue();
```
Comment (leoromanovsky) [inline src/main/java/com/eppo/sdk/EppoClient.java]:
```suggestion
            String variationValue = assignmentResult.getVariation().getTypedValue().toString()
```

`stringValue` assumes a "string` variation type right? Perhaps you meant to call `toString` on this to handle any type?

File: src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java [hunk 12]
```diff
@@ -28,45 +25,32 @@ public static List<Variation> evaluateBanditVariations(
...
-            // Use the first 4 bytes to create an int
-            int digestInt = ByteBuffer.wrap(digest, 0, 4).getInt();
-            // Ensure its positive
-            return Math.abs(digestInt);
-        } catch (NoSuchAlgorithmException e) {
-            throw new RuntimeException("Could not generate MD5", e);
-        }
+    private static int hashToPositiveInt(String experimentKey, String subjectKey, String actionKey) {
+        int SHUFFLE_SHARDS = 10000;
+        return Shard.getShard(experimentKey+"-"+subjectKey+"-"+actionKey, SHUFFLE_SHARDS);
```
Comment (leoromanovsky) [inline src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java:42]: nice

**Review by leoromanovsky — APPROVED** [review]

File: src/main/java/com/eppo/sdk/EppoClient.java (hunk 10 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/EppoClient.java:99]: This is within `getAssignedVariation()` now with the end goal to return a `VariationAssignmentResult`. I use `null` to basically indicate no assignment could be made (vs. always returning a result but with a null variation)

File: src/main/java/com/eppo/sdk/EppoClient.java (hunk 11 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/EppoClient.java]: This is a good call out! For now, we'll have all bandit experiments use string types only. But I can change to use `toString()` to help set us well for a future where that may not be the case.

File: src/main/java/com/eppo/sdk/helpers/Shard.java (hunk 8 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/helpers/Shard.java:47]: Ha yeah. In the former way, the compiler would probably just optimize to this anyways. But doing this appeased the suggestion from WebStorm 😅

File: src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java (hunk 12 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/helpers/bandit/BanditEvaluator.java:42]: Code reuse FTW!

File: src/test/resources/bandits/rac-experiments-bandits-beta.json (hunk 9 above)
Comment (aarsilv) [reply src/test/resources/bandits/rac-experiments-bandits-beta.json:35]: I wasn't planning on it (and don't depend on it--the common tests still pass) but could if you think that's best!

File: src/main/java/com/eppo/sdk/dto/AssignmentLogData.java [hunk 13]
```diff
@@ -8,34 +8,24 @@
 public class AssignmentLogData {
     public String experiment;
     public String featureFlag;
-    public String assignmentModelVersion;
```
Comment (sameerank) [inline src/main/java/com/eppo/sdk/dto/AssignmentLogData.java:11]: Backtracking on changes from https://github.com/Eppo-exp/java-server-sdk-bandit-beta/pull/1 to put this in bandit log data

File: src/main/java/com/eppo/sdk/EppoClient.java [hunk 14]
```diff
@@ -68,96 +51,153 @@ private Optional<EppoValue> getAssignmentValue(
...
+
+        Variation assignedVariation = assignmentResult.getVariation();
+        Optional<EppoValue> assignmentValue = Optional.of(assignedVariation.getTypedValue());
+
+        // Below is used for logging
+        String experimentKey = assignmentResult.getExperimentKey();
+        String allocationKey = assignmentResult.getAllocationKey();
+        String assignedVariationString = assignedVariation.getTypedValue().stringValue();
+
+        if (assignedVariation.getAlgorithmType() == AlgorithmType.OVERRIDE) {
```
Comment (sameerank) [inline src/main/java/com/eppo/sdk/EppoClient.java:68]: Not really familiar with Java. Are these get functions generated by lombok?

**Review by sameerank — APPROVED** [review]

File: src/main/java/com/eppo/sdk/EppoClient.java (hunk 14 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/EppoClient.java:68]: They are. I'm not a fan of lombok automagic--even if it does save keystrokes--but it was already in this codebase, so I'm rolling with it. Happy to have us rip it out in a later PR 😅

File: src/main/java/com/eppo/sdk/dto/AssignmentLogData.java (hunk 13 above)
Comment (aarsilv) [reply src/main/java/com/eppo/sdk/dto/AssignmentLogData.java:11]: Correct (and nice job spotting that)! The bandits team is currently leaning towards leaving the existing assignments log table alone and putting bandit stuff in specifically bandit-purposed tables. ([relevant design doc section](https://www.notion.so/eppo/Bandits-Engineering-Design-c5df5b6072b446aa8cacbf54e4d1f9df?pvs=4#263ada470ab34b8c91a906914788dd3f))

Note we'll still log the variation (e.g., `bandit`) to assignments when the bandit is selected, but it's action (e.g., `nike ad`) will be logged to the bandits table along with the variation and other useful metadata. Some data will be duplicated, but data is "cheap" 😅
